### PR TITLE
[in_app_purchase][iOS] fix iOS promotional offers (SKPaymentDiscountWrapper was not used properly)

### PR DESCRIPTION
**Pull request by crystalstorm**
Make iOS Promotional Offers in in_app_purchase plugin usable. `SKPaymentDiscountWrapper` was not being sent trough the method channel.
This is an update to an old (already closed) [PR here](https://github.com/flutter/plugins/pull/6099)

Resolves: [[in-app-purchase][iOS] Promotional Offers in iOS not working](https://github.com/flutter/flutter/issues/107533#issue-1303299545)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/FIAObjectTranslator.m
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/FIAObjectTranslator.m:280]: Nice catch. Can we add an XCUnitTest for this? Maybe have the timestamp to be a very big long long number?

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:409]: We should have some test for the `toMap` method, can we update the test to include the discount? If we don't have an existing test, do you mind adding a test for the `toMap` method?

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:79]: We should test that the `paymentDiscount` is encoded and sent to method channel.

**Review by cyanglaz — CHANGES_REQUESTED** [review]

File: packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/FIAObjectTranslator.m
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/FIAObjectTranslator.m:280]: Done. I've added XCUnitTest using current timestamp number value. It's enough to cause the `intValue` overflow in negative number. If I use really big long long number, the intValue overflow goes positive again and passes the test. I'm not sure of the usefulness of this "not negative int" check in the original code.

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:409]: Done. I've added a new test checking the values of the `SKPaymentWrapper` object and the result of the `toMap` method explicitly.

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:79]: I see your point, that in fact the bug was that the discount parameter was not sent trough the method channel. But I am not sure on how to construct such a test?
In my understanding: now the discount is send trough `AppStorePurchaseParam` in the `buyNonConsumable` method, which is not returning a useful result to test. This method constructs `SKPaymentWrapper` with a discount parameter and sends it to `_skPaymentQueueWrapper.addPayment` which turns the `SKPaymentWrapper` to a map (which we test in the newly created test for the toMap method) and sends it trough the method channel. But the channel is not returning useful result for testing either. The encoding of the `paymentDiscount` is in the `toMap` method which we test now. 
Can you elaborate please?

Comment (cyanglaz) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:79]: You might be able to intercept the method call in the `FakeStoreKitPlatform` and inspect the value?

File: packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/TranslatorTests.m
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/TranslatorTests.m:406]: Can we also check members in the `paymentDiscount` have the same value with `discountMap`. In the case of this Pr, especially `timestamp`

Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/example/ios/RunnerTests/TranslatorTests.m:406]: Done.

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/lib/src/in_app_purchase_storekit_platform.dart:79]: I got your idea. Thanks! Done in the last commit.

**Comment by crystalstorm** [comment]
@cyanglaz I don't understand while the last commit failed at the `"submit-queue"` test. When I read the details it seems related to android. I cannot see reference to the changes made in the PR here. Can you help?

**Comment by cyanglaz** [comment]
> @cyanglaz I don't understand while the last commit failed at the `"submit-queue"` test. When I read the details it seems related to android. I cannot see reference to the changes made in the PR here. Can you help?

"submit-queue" is not related to this PR. It is an indication of the current tree's status. This CI is intended to block merging when tree is red (when the build is failing on master). Usually, if there is a "submit-queue" error, someone on the flutter team is working on a fix, the best action is to wait.

File: packages/in_app_purchase/in_app_purchase_storekit/lib/src/types/app_store_purchase_param.dart
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_storekit/lib/src/types/app_store_purchase_param.dart:37]: nits, the "optional" keyword is probably not necessary as the nullability of this parameter already indicates it is optional.
However, it might be better to add an explanation what does it mean to be null, something like:
```
///
/// The value is `null` when the product does not have a discount.
```

File: packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart
Comment (cyanglaz) [inline packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart:33]: nits: Alternatively, we can make it non-nullable and initialized it to empty map, this can avoid many unwrapping.

**Review by cyanglaz — APPROVED** [review]
LGTM mod nits, @stuartmorgan could you do a secondary review?

File: packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart:182]: This comment sounds like you attempted to cast with `as` instead of `cast<...>`.

File: packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart
Comment (stuartmorgan) [inline packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart:180]: Why is this `dynamic` rather than `Object?`?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/in_app_purchase/in_app_purchase_storekit/pubspec.yaml
Comment (cyanglaz) [reply packages/in_app_purchase/in_app_purchase_storekit/pubspec.yaml:5]: This adds a new field to the `AppStorePurchaseParam` so it should be a new feature. The PR description or CHANGELOG should probably be something like "Supports adding discount information to `AppStorePurchaseParam`", instead of "fixing".

Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase_storekit/pubspec.yaml:5]: Thanks for the correction there; I didn't looks closely enough at the details and thought it was just populating something that was previously added but not wired up.

File: packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart]: Done. Changed in the last commit.

File: packages/in_app_purchase/in_app_purchase_storekit/CHANGELOG.md
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/CHANGELOG.md]: Changed in the last commit.

File: packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart]: Changed in the last commit

File: packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/test/fakes/fake_storekit_platform.dart]: True. Changed it to a `cast<>` now. Thanks!

File: packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/FIAObjectTranslator.m
Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/FIAObjectTranslator.m:240]: > when I was buying a product without discount, it was `NULL`

You mean it was `NSNull`, presumably, given the code you added. `NULL` is a typedef for `0`, and is essentially the C equivalent of `nil`. `NSNull` is not `nil`/`NULL`, it is singleton class.

> but it did not stop at the `!map`

That is expected; a pointer that is pointing to an `NSNull` instance is not `nil`.

> it went to the next check `map.count <= 0` and crashed.

Presumably the crash was an `unrecognized selector sent to instance` exception, because `count` is not a method of `NSNull`.

> So here I do not try to check for a type violation

Yes, that is exactly what you are checking for. You are taking a pointer of type `NSDictionary` and checking whether it contains an `NSNull` instance instead of an `NSDictionary` instance. Which would be a type violation.

> As for if it should be done here or in the caller. Why it is better in the caller?

This method takes an `NSDictionary*`, not an `id`. It is better in the caller because it is a type violation to pass an argument to this method that is not an `NSDictionary`, and an instance of `NSNull` **is not** an `NSDictionary`.

The comparable Dart code is:

```
class NSNull {
}

void methodHandler(dynamic args) {
  expectsAMap(args);
}

void expectsAMap(Map? map) {
  if (map != null) {
    print(map.length);
  }
}

void main() {
  NSNull nullArg = NSNull();
  methodHandler(nullArg);
}
```

The bug is this Dart code is `methodHandler` passing `args` to `expectsAMap` without checking whether it's a `Map`, not when `expectsAMap` uses `map` as if it were its declared type.

> Shouldn't the method take care of the argument's validation, as it does in the checks that come below?

Methods should not have to validate that they haven't been called with types that violate their declarations. Other validation (like whether it's `nil`), sure.

The code below this is also wrong; values with unknown types should be type checked before assigning to a variable declared as a specific type, not after. Either the values can only be `NSString`s, in which case the checks are pointless, or (more likely) they could be `NSString`s or `NSNull`s, in which case the variables should be declared `id`, not `NSString*`.

**Comment by crystalstorm** [comment]
@cyanglaz / @stuartmorgan should I squash all the commits before you accept the changes and decide to merge?

**Comment by jmagman** [comment]
> @cyanglaz / @stuartmorgan should I squash all the commits before you accept the changes and decide to merge?

That's not necessary, GitHub will squash when it merges.

File: packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/InAppPurchasePlugin.m
Comment (stuartmorgan) [reply packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/InAppPurchasePlugin.m:203]: (@cyanglaz Any chance you could plan to convert this plugin to Pigeon sometime in the nearish term, like early 2023? It seems like there's a fair amount of this kind of problem in the plugin, and we could eliminate all of it with a Pigeon conversion.)

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Looks good overall, just one more type issue.

File: packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/InAppPurchasePlugin.m
Comment (crystalstorm) [reply packages/in_app_purchase/in_app_purchase_storekit/ios/Classes/InAppPurchasePlugin.m]: Done. I did it with ObjC version of the helper that you've shown, one line without the condition check. Thanks! :)

**Review by stuartmorgan — APPROVED** [review]
LGTM, thanks for the iteration on the type issues!

**Pull request merged**
